src/dsc.h

```diff
@@ -417,46 +441,38 @@ class IdList {
         AllocForOneMore();
 
         // Copy-construct at the end of the list.
+        indexes[n] = n;
         new(&elem[n]) T(*t);
         ++n;
 
         return t->h;
     }
 
-    T * LowerBound(T const& t) {
-        if(IsEmpty()) {
-            return nullptr;
-        }
-        auto it = std::lower_bound(begin(), end(), t, Compare());
-        return it;
+    Iterator LowerBound(T const& t) {
+        return std::lower_bound(begin(), end(), t, Compare());
     }
 
-    T * LowerBound(H const& h) {
-        if(IsEmpty()) {
-            return nullptr;
-        }
-        auto it = std::lower_bound(begin(), end(), h, Compare());
-        return it;
+    Iterator LowerBound(H const& h) {
+        return std::lower_bound(begin(), end(), h, Compare());
     }
 
     int LowerBoundIndex(T const& t) {
-        if(IsEmpty()) {
-            return 0;
-        }
-        auto it = LowerBound(t);
-        auto idx = std::distance(begin(), it);
-        auto i = static_cast<int>(idx);
-        return i;
+        return std::distance(begin(), LowerBound(t));
     }
+
     void ReserveMore(int howMuch) {
         if(n + howMuch > elemsAllocated) {
             elemsAllocated = n + howMuch;
+            int *newIndexes = (int *)::operator new[]((size_t)elemsAllocated*sizeof(int));
```

Comment:
[Using `vector` for the storage](https://github.com/ruevs/solvespace/blob/6bc1aa91c6a0f1acfc986b34ff54ed36efcdbfc1/src/dsc.h#L404) gets rid of the manual memory management, and hopefully at least in release mode should be as fast.

Reply:
That's one of the things that should be unified throughout the codebase, I'm just using what's already there.

I don't even know how the custom allocator library is hooked up.

@Evil-Spirit wanted to use `realloc`, which should be a trick that `std::vector` could be able to do as well, though I'm dubious whether it's useful for geometric steps.

Reply:
The important bit is that idlist is reference semantics, not value semantics, so you can't embed a std::vector, but rather a std::shared_ptr<std::vector> - a lesson I learned painfully.

Reply:
@rpavlik Do you mean to say that once a pointer to an element is "taken out' of an IdList it may be kept around and dereferenced after the list has been modified? And thus the pointer may be invalid?

If this is the case can you point to an example?

But this does not make sense - the original implementation of IdList did not guarantee that the elemnts will not move - in fact it moved them around all the time!

In fact my implementation is "better" in this respect since it never moves elements once they are added because the `elemstore` is not rearranged when elements are removed:
https://github.com/ruevs/solvespace/blob/3c6e2be4b920d706d8231030467c5271a6d87d85/src/dsc.h#L570
the removed elements are simply added to the `freelist`.